### PR TITLE
GAP-2015: Adds date created to super admin dashboard

### PR DESCRIPTION
**Pull request by kiramarstonTCO**

### CODE REVIEW

File: src/main/java/gov/cabinetofice/gapuserservice/dto/UserDto.java [hunk 1]
```diff
@@ -15,6 +20,8 @@ public class UserDto {
     private List<Role> roles;
     private Role role;
     private Department department;
+    private Optional<Instant> created;
```
Comment (john-tco) [inline src/main/java/gov/cabinetofice/gapuserservice/dto/UserDto.java:23]: Will this date be the moment the user is created? Do we instead want to set this value when the user runs through OL `/login`

File: src/main/java/gov/cabinetofice/gapuserservice/model/User.java [hunk 2]
```diff
@@ -38,6 +41,14 @@ public class User {
     @Enumerated(EnumType.STRING)
     private LoginJourneyState loginJourneyState;
 
+    @CreatedDate
+    @Column(name = "created")
+    private Instant created;
```
Comment (john-tco) [inline src/main/java/gov/cabinetofice/gapuserservice/model/User.java:46]: Will this column populate when the user row is created? Do we instead want to set this value when the user runs through OL `/login`

File: src/main/java/gov/cabinetofice/gapuserservice/dto/UserDto.java [hunk 3]
```diff
@@ -5,7 +5,12 @@
 import gov.cabinetofice.gapuserservice.model.User;
 import lombok.Data;
 
+import java.time.Instant;
+import java.time.ZoneId;
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/dto/UserDto.java:9]: Some of these imports are unused so can be removed

File: src/main/java/gov/cabinetofice/gapuserservice/model/User.java (hunk 2 above)
Comment (kiramarstonTCO) [reply src/main/java/gov/cabinetofice/gapuserservice/model/User.java:46]: Yes, I spoke to Dylan about this one: 
Screenshot 2023-08-18 at 15 04 59

File: src/main/java/gov/cabinetofice/gapuserservice/dto/UserDto.java (hunk 1 above)
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/dto/UserDto.java:20]: We don't want to use Optionals as a field type, instead we can make them @Nullable

```
@Nullable
    private Instant created;
    @Nullable
    private Instant updated;

    public UserDto(final User user) {
        this.gapUserId = user.getGapUserId().toString();
        this.emailAddress = user.getEmailAddress();
        this.sub = user.getSub();
        this.department = user.getDepartment();
        this.roles = user.getRoles();
        this.role = user.getHighestRole();
        this.created = user.getCreated();
        this.updated = user.getUpdated();
    }
```

File: src/main/java/gov/cabinetofice/gapuserservice/model/User.java (hunk 2 above)
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetofice/gapuserservice/model/User.java:46]: I believe @CreatedDate will automatically only create the date when a user resource is added.

I can see that we are saving the user again after the user logs in via One login

` public User createOrGetUserFromInfo(final OneLoginUserInfoDto userInfo) {
        final Optional<User> userOptional = getUserFromSubOrEmail(userInfo.getSub(), userInfo.getEmailAddress());
        if (userOptional.isPresent()) {
            final User user = userOptional.get();
            if (!user.hasSub()) {
                user.setSub(userInfo.getSub());
                return userRepository.save(user);
            }
            return user;
        }
        return createNewUser(userInfo.getSub(), userInfo.getEmailAddress());
    }`
    
    
    Can you confirm this is adds the dates automatically @kiramarstonTCO ?

**Review by ryan-tco — APPROVED** [review]

**Pull request merged**